src/ACL.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+namespace atk4\login;
+
+use atk4\core\Exception;
+use atk4\data\Model;
+use atk4\data\Persistence;
+
+/**
+ * Access Control Layer. Create one and pass it to your Auth controller.
+ */
+class ACL
+{
+    /**
+     * References an auth controller, so we can look up who is logged
+     * in and what their permissions are.
+     *
+     * @var Auth
+     */
+    public $auth;
+
+    /**
+     * Returns AccessRules model for logged in user and in model scope.
+     *
+     * @param Model $model
+     *
+     * @return \atk4\login\Model\AccessRule
+     */
+    public function getRules(Model $model)
+    {
+        /** @var \atk4\login\Model\User*/
+        $user = $this->auth->user;
+
+        if (!$user->loaded()) {
+            // user is not logged in - let's force him to do so. Alternative is to throw exception, but that's ugly.
+            $this->auth->check();
+            //throw new Exception('User should be logged in!');
+        }
+
+        return $user->ref('AccessRules')->addCondition('model', get_class($model));
+    }
+
+    /**
+     * Given a model, this will apply some restrictions on it.
+     *
+     * Extend this method if you wish.
+     *
+     * @param Persistence $p
+     * @param Model       $m
+     */
+    public function applyRestrictions(Persistence $p, Model $m)
+    {
+        foreach ($this->getRules($m) as $rule) {
+
+            // set visible and editable fields
+            foreach ($m->getFields() as $name => $field) {
+                $field['ui']['visible'] = $rule['all_visible'] || (array_search($name, $rule['visible_fields']) !== false);
+                $field['ui']['editable'] = $rule['all_editable'] || (array_search($name, $rule['editable_fields']) !== false);
+            }
+
+            // remove not allowed actions
+            if (!$rule['all_actions'] && $rule['actions']) {
+                $actions_to_remove = array_diff(array_keys($m->getActions()), $rule['actions']);
+                foreach ($actions_to_remove as $action) {
+                    $m->getAction($action)->enabled = false;
+                }
+            }
+
+            // add conditions on model
+            if ($rule['conditions']) {
+                $this->applyConditions($p, $m, $rule['conditions']);
```

Comment:
is this works? how?

Reply:
It will in future when we will have nice UI for defining conditions and $rule['conditions'] will contain json encoded structure of conditions to apply on model. For now I guess I will have to comment it out.